task-maker-async/src/store.rs

```diff
@@ -43,58 +82,591 @@ pub enum FileReadingOutcome {
-
-    /// Opens a file for reading. Creates a lease for the input fileset that will prevent it
-    /// from being dropped. If the file is not present and finalized, returns an error immediately.
-    async fn open_input_file(hash: InputFileHash) -> Result<FileSetHandle, ()>;
+    /// Used by the client to upload inputs to the DAG. Returns a writing handle if the input file is
+    /// not present, or a reading handle if it is.
+    async fn create_or_open_input_file(hash: InputFileHash) -> Result<FileSetHandle, Error>;
 
     // Creating a computation is not a RPC.
 
     /// Opens a computed fileset for reading. Creates a lease for the computation data that will prevent it
     /// from being dropped. If the computation is not present, waits until it is created.
-    async fn open_computation(hash: ComputationHash) -> FileSetHandle;
+    async fn open_computation(hash: ComputationHash) -> Result<FileSetHandle, Error>;
 
     /// Opens a file inside a fileset. Waits for the file to be created if it doesn't exist yet and
     /// the fileset handle is a reading handle, creates the file otherwise.
     /// Returns an error if the handle is invalid.
-    async fn open_file(handle: FileSetHandle, file: FileSetFile) -> Result<FileHandle, ()>;
+    async fn open_file(handle: FileSetHandle, file: FileSetFile) -> Result<FileHandle, Error>;
 
     /// Appends data to a file in a fileset that is open for writing. Refreshes the writing lease.
-    async fn append_chunk(file: FileHandle, data: Vec<u8>) -> Result<(), ()>;
+    async fn append_chunk(file: FileHandle, data: Vec<u8>) -> Result<(), Error>;
 
     /// Finalizes a FileSet handle in writing mode. Terminates the writing lease and returns a reading
     /// lease for the same FileSet. If finalizing an input fileset, returns an error if the hash of
     /// its MainFile is not correct.
-    async fn finalize_fileset(handle: FileSetHandle) -> Result<FileSetHandle, ()>;
+    async fn finalize_fileset(handle: FileSetHandle) -> Result<FileSetHandle, Error>;
 
     /// Tries to read from a file. Refreshes the corresponding lease.
-    async fn read_chunk(file: FileHandle, offset: usize) -> FileReadingOutcome;
+    async fn read_chunk(file: FileHandle, offset: usize) -> Result<FileReadingOutcome, Error>;
 
     /// Refreshes the lease for the given fileset.
     /// It is guaranteed that the fileset will not be deleted while there's an outstanding lease
     /// to it. It is an error to refresh a lease of a non-existent input.
-    async fn refresh_fileset_lease(handle: FileSetHandle) -> Result<(), ()>;
+    async fn refresh_fileset_lease(handle: FileSetHandle) -> Result<(), Error>;
+}
+
+#[derive(Debug, Eq, PartialEq)]
+enum FileSetKind {
+    Computation,
+    InputFile,
+}
+
+type FileContents = Vec<u8>;
+
+#[derive(Debug)]
+struct FileSet {
+    files: HashMap<FileSetFile, FileContents>,
+    kind: FileSetKind,
+    finalized: bool,
 }
 
+type FileSetVariants = HashMap<HashData, FileSet>;
```

Comment:
What is this `HashData` an hash of? If possible, try to do other type aliases also for the hashed _things_. Here and below at `StoreServiceImpl::filesets` it's not clear what they're hashes of.